python/cuml/ts/arima.pyx

```diff
@@ -0,0 +1,814 @@
+            dx = xs[1] - xs[0]
+            xfc = np.linspace(1, 1+50*dx, 50)
+            plt.plot(xs, ys, xfc, yfc)
+
+        """
+
+        p, d, q = self.order[0]
+        
+        handle = cuml.common.handle.Handle()
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>handle.getHandle()
+
+        x = pack(p, d, q, self.num_batches, self.mu, self.ar_params, self.ma_params)
+        cdef uintptr_t d_params_ptr
+        d_params, d_params_ptr, _, _, _ = input_to_dev_array(x, check_dtype=np.float64)
+
+        cdef np.ndarray[double, ndim=2, mode="fortran"] y_p = np.zeros(((self.num_samples),
+                                                                        self.num_batches), order="F")
+        cdef uintptr_t d_y_p_ptr
+        d_y_p, d_y_p_ptr, _, _, _ = input_to_dev_array(y_p, check_dtype=np.float64)
+
+        cdef np.ndarray[double, ndim=2, mode="fortran"] vs = np.zeros(((self.num_samples - d),
+                                                                       self.num_batches), order="F")
+        cdef uintptr_t d_vs_ptr
+        d_vs, d_vs_ptr, _, _, _ = input_to_dev_array(vs, check_dtype=np.float64)
+
+        cdef uintptr_t d_y_ptr    
+        if self.d_y is None:
+            d_y, d_y_ptr, _, _, _ = input_to_dev_array(self.y, check_dtype=np.float64)
+            self.d_y = (d_y, d_y_ptr)
+            
+        (_, d_y_ptr) = self.d_y
+
+        residual(handle_[0], <double*>d_y_ptr, self.num_batches, self.num_samples, p, d, q,
+                 <double*>d_params_ptr, <double*>d_vs_ptr,
+                 False)
+
+        y_diff = np.diff(self.y, axis=0)
+        cdef uintptr_t d_y_diff_ptr
+        d_y_diff, d_y_diff_ptr, _, _, _ = input_to_dev_array(y_diff, check_dtype=np.float64)
+
+        cdef np.ndarray[double, ndim=2, mode="fortran"] y_fc = np.zeros((nsteps, self.num_batches), order="F")
+        cdef uintptr_t d_y_fc_ptr
+        d_y_fc, d_y_fc_ptr, _, _, _ = input_to_dev_array(y_fc, check_dtype=np.float64)
+
+        forecast(handle_[0], nsteps, p, d, q, self.num_batches, self.num_samples, <double*> d_y_ptr,
+                 <double*>d_y_diff_ptr, <double*>d_vs_ptr, <double*>d_params_ptr, <double*> d_y_fc_ptr)
+
+        updateHost(&y_fc[0,0], <double*>d_y_fc_ptr, nsteps * self.num_batches, 0)
+
+        return y_fc
+
+
+def estimate_x0(order: Tuple[int, int, int],
+                nb: int,
+                yb) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Provide initial estimates to ARIMA parameters `mu`, `ar`, and `ma` for the batched input `yb`"""
+    pynvtx_range_push("init x0")
+    (p, d, q) = order
+    N = p + d + q
+    x0 = np.zeros(N * nb)
+
+    for ib in range(nb):
+        y = yb[:, ib]
+
+        if d == 1:
+            yd = np.diff(y)
+        else:
+            yd = np.copy(y)
+
+        x0ib = _start_params((p, q, d), yd)
+
+        x0[ib*N:(ib+1)*N] = x0ib
+
+    mu, ar, ma = unpack(p, d, q, nb, x0)
+
+    pynvtx_range_pop()
+
+    return mu, ar, ma
```

Comment:
This doesn't call C++, right? Could this be a numba cuda kernel?

Reply:
This is one of the methods I never ported as `_start_params` had a high level of complexity and used several numpy primitives (`np.linalg.lstsq`) which were daunting to port.